python/bin/watchman-wait-aio

```diff
@@ -0,0 +1,241 @@
+#!/usr/bin/env python3
+
+from pywatchman.aioclient import AIOClient as WatchmanClient
+
+import argparse
+import asyncio
+import pywatchman
+import os
+import sys
+
+def _with_timeout(task, timeout):
+    if timeout > 0:
+        return asyncio.wait_for(task, timeout)
+    return task
+
+
+class Subscription(object):
+    '''A single subscription.'''
+
+    def __init__(self, abspath, name):
+        self.name    = name    # Our name for this subscription
+        self.abspath = abspath # Full path to the dir being watched
+        self.root    = None    # Watched root
+        self.relpath = None    # Path relative to root of thr dir being watched
+
+    def __str__(self):
+        return 'Subscription(name=%s, root=%s, abspath=%s, relpath=%s)' % (
+            self.name, self.root, self.abspath, self.relpath)
+
+    async def activate(self, client, patterns=None, fields=['name']):
+        '''Activate the subscription. This function is a coroutine.'''
+        dir_to_watch = self.abspath
+
+        if patterns:
+            expr = ['anyof']
+            for p in patterns:
+                expr.append(['match', p, 'wholename', {
+                    'includedotfiles': True}])
+        else:
+            expr = ['true']
+
+        if not os.path.isdir(self.abspath):
+            # Need to watch its parent
+            dir_to_watch = os.path.dirname(self.abspath)
+            expr = ['name', os.path.basename(self.abspath)]
+
+        query = {
+            'expression': expr,
+            'fields': fields,
+        }
+
+        watch = await client.query('watch-project', dir_to_watch)
+
+        if 'warning' in watch:
+            print('WARNING: ', watch['warning'], file=sys.stderr)
+
+        self.root = watch['watch']
+        if 'relative_path' in watch:
+            self.relpath = watch['relative_path']
+            expr = ['allof', ['dirname', self.relpath], expr]
+
+        # get the initial clock value so that we only get updates
+        query['since'] = (await client.query('clock', self.root))['clock']
+
+        await client.query('subscribe', self.root, self.name, query)
+
+    async def get(self, client):
+        data = await client.get_subscription(self.name, self.root)
+        return data['files']
+
+    @classmethod
+    def from_path(cls, path, name=None):
+        '''Create a subscription for the given path, using that path as the
+        subscription's name if not specified.'''
+        name = name if name is not None else path
```

Comment:
nit: not important for now but in Python you can do `name = name or path` which matches your case well enough and is easier to read :)